### PR TITLE
fix(moe): validate expert capacity in DeepSeek fused routing

### LINKED ISSUE #4867: [CI] Invalid top-k grouping in GLM-5 and Kimi-K2.5 NVFP4 MoE benchmarks
## Problem

The `trtllm_fp4_block_scale_moe` benchmark cases `GLM-5-NVFP4` and `Kimi-K2.5-NVFP4` fail before producing benchmark results on B300 and GB300.

The cases pass `--top_k 8 --n_group 1 --topk_group 1`, which violates the benchmark's grouping constraint:

```text
Invalid configuration: topk_group * n_group (1) must be >= topk (8) and topk_group (1) must be <= n_group (1)
```

The same failure was reproduced in completed runs for commits `a29ed04`, `2f52955`, `2cc51dc`, `d981061`, and `faf7c6a`.

Evidence: [B300 job](https://nv/flashinfer-ci/-/jobs/419258296), [GB300 job](https://nv/flashinfer-ci/-/jobs/419258295).

## Reproduction

```bash
python3 flashinfer_benchmark.py --routine trtllm_fp4_block_scale_moe --num_tokens 64 --hidden_size 6144 --intermediate_size 2048 --num_experts 256 --top_k 8 --n_group 1 --topk_group 1 --routed_scaling_factor 2.5 --use_routing_bias --routing_method deepseek_v3 --use_shuffled_weight --activation-type Swiglu --fp4_mode nvfp4 --verbose --generate_repro_command --case_tag GLM-5-NVFP4

python3 flashinfer_benchmark.py --routine trtllm_fp4_block_scale_moe --num_tokens 1 --hidden_size 7168 --intermediate_size 2048 --num_experts 384 --top_k 8 --n_group 1 --topk_group 1 --routed_scaling_factor 2.827 --use_routing_bias --routing_method deepseek_v3 --use_shuffled_weight --activation-type Swiglu --fp4_mode nvfp4 --verbose --generate_repro_command --case_tag Kimi-K2.5-NVFP4
```

## Expected behavior

These benchmark cases should use a valid group configuration (or omit group-limited routing) and produce benchmark results.

## Actual behavior

Both cases are rejected by input validation, so no performance result is generated.

### PR DESCRIPTION
**Pull request by time-machine2025**
## 📌 Description

`fused_topk_deepseek` rejects the GLM-5 and Kimi-K2.5 benchmark routing configurations (`n_group=topk_group=1`, `topk=8`, 256/384 experts) before kernel execution. The check counts groups instead of the experts available in the selected groups. Validate against `topk_group * (num_experts // n_group)`, including equality, and reject insufficient capacity.

For the newly admitted default single-group top-k calls, mask unwritten slots in the intermediate shared-memory reduction. Otherwise, top-k below eight can select stale candidates. Keep Cake's single-group top-1 limitation explicit in both its Python and native checks; its grouped checks use the corrected expert capacity. Also align the default grouped checks with the kernel's limits: at most eight groups, at most four selected groups, and at least two experts per group. Validate positive group counts before native division and reject invalid top-k even when the Python check is bypassed. Update the API documentation and the numerical/replay tests' matching skip conditions.

## 🔍 Related Issues

Fixes #4867

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.). Full GPU CI matrix requires upstream CI; completed validation is listed below.

Validated against main `715c3c9f15af2ffbf14068fabcb1f39af19c47dd`, using repository-pinned CUTLASS/CCCL/spdlog dependencies:

- The updated CPU validation file passes with no visible GPU: **62 passed**. Against the previous PR head (`0a17574`), the new grouped-limit and zero-top-k regressions produce **4 failed / 58 passed**. The original capacity checks also reproduced the issue on upstream baseline (9 failed / 46 passed).
- Both affected test files on an NVIDIA L40S, Python 3.12.14, PyTorch 2.13.0+cu130: **956 passed, 3960 skipped**. Skips cover unsupported parameter combinations and Cake's SM100/SM103 requirement. New tests require the expected diagnostic for invalid inputs and compare exact expert IDs plus per-expert weights at legal boundaries.
- An independent CPU boundary sweep checks **42,213 configurations** against a separately constructed parameter contract: all agree after the fix; the previous PR head incorrectly accepts 2,075.
- An independent GPU sweep passes **216 legal shape/dtype combinations**, with exact expert IDs, per-expert weight checks, and replay-buffer checks. It also verifies **48 native rejections** across direct FFI and `skip_check=True` calls, unchanged output sentinels, and a successful legal call after every rejection. Compute Sanitizer **memcheck and synccheck both report 0 errors** on this sweep.
- The original **198-configuration** public-API sweep still passes, including a separate CUDA stream and three CUDA Graph replays per configuration. The original 24 negative-bias reuse runs passed, and seven top-8 controls were bitwise identical between baseline and the initial fix.
- Full `pre-commit run --all-files`: **14 passed, 1 skipped**.
- Cake was freshly compiled for SM100a and loaded, without launching its kernels.

Commands:
```bash
python -m pytest tests/model_optimizations/test_dsv3_fused_routing.py \
  tests/model_optimizations/test_dsv3_fused_routing_backend_selection.py -q
pre-commit run --all-files
```

The full B300/GB300 NVFP4 MoE benchmarks and Cake GPU numerics were not run.

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

AI-assisted implementation and review. The change keeps the benchmark's model routing parameters intact and does not change the routing algorithm or claim a performance improvement. Particular review points are the inclusive capacity boundary, agreement between Python and native scalar checks, and preserving Cake's actual single-group top-1 contract.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved fused routing reliability by preventing invalid intermediate expert selections during multi-warp processing.
  - Strengthened validation for grouped routing, including expert capacity, group counts, and supported top-k ranges.
  - Improved backend selection for single-group and multi-group routing configurations.
  - Unsupported configurations are now rejected earlier with clearer validation behavior.

- **Tests**
  - Added coverage for routing edge cases, backend selection, warmup reuse, invalid parameters, and boundary-capacity configurations.

### CODE REVIEW

File: csrc/fused_moe/noAuxTcKernels.cu [hunk 1]
```diff
@@ -263,13 +270,16 @@ void invokeNoAuxTc(InputT* scores, BiasT* bias, OutputT* topk_values, IdxT* topk
...
+  // ``MaxNumTopGroups`` selected groups, and needs >= 2 experts per group.
+  int64_t const experts_per_group = n_group > 0 ? num_experts / n_group : 0;  // 0 => unsupported
 
-  int64_t const experts_per_group = num_experts / n_group;
-  bool const is_multi_group = (n_group != 1) && (num_experts <= NumDeepseekExperts) &&
-                              (experts_per_group <= WARP_SIZE) &&
-                              (experts_per_group * topk_group <= MaxNumExpertsUnit);
+  bool const is_single_group = (n_group == 1) && (num_experts <= NumKimiK2Experts);
+  bool const is_multi_group =
+      (n_group > 1) && (n_group <= MaxNumGroupScores) && (num_experts <= NumDeepseekExperts) &&
```
[the remark this answers is not in the document]
Comment (time-machine2025) [reply csrc/fused_moe/noAuxTcKernels.cu:280]: Could you re-check this finding against CUDA's rule for exited threads? The [CUDA 13.0 Programming Guide](https://docs.nvidia.com/cuda/archive/13.0.0/cuda-c-programming-guide/index.html#independent-thread-scheduling-7-x) states that `__syncthreads()` waits for “all non-exited threads in the block.”

Here, `warpIdx >= numGroup` returns entire warps, while every remaining thread reaches the same barrier. For `num_experts=64, n_group=2, topk_group=1, topk=1`, warps 0 and 1 write all 64 expert entries and group-score slots 0 and 1. Subsequent reads are restricted to those active groups and their valid expert entries; they do not require writes from the exited warps.

I also tested this exact configuration on the unchanged `ea92583` source using an L40S (SM89), CUDA 13.0, and PyTorch 2.13.0+cu130. The sweep covered all nine FP32/FP16/BF16 score/bias dtype pairs, token counts 1/7/64, and both values of `launch_with_pdl`: 54 cases, each with 10 direct calls and 3 CUDA Graph replays. Expert IDs matched exactly, per-expert weights met the dtype-specific tolerances, and replay outputs and output guard rows matched expectations, including inputs where either active group wins. Compute Sanitizer `synccheck` and `memcheck` each reported 0 errors. This hardware result is supporting evidence, not a claim that every supported GPU was tested.

Could you identify a non-exited thread that skips the barrier, or a shared-memory read that depends on an exited warp's write? That would help clarify whether there is an additional constraint I am missing.

File: flashinfer/fused_moe/fused_routing_dsv3.py [hunk 2]
```diff
@@ -141,11 +148,29 @@ def _check_dsv3_fused_routing_supported(
...
+    if num_experts % n_group != 0:
+        raise ValueError(
+            f"Invalid configuration: num_experts ({num_experts}) must be divisible by "
+            f"n_group ({n_group})"
+        )
+    if topk_group <= 0 or topk_group > n_group:
+        raise ValueError(
+            f"Invalid configuration: topk_group ({topk_group}) must be in [1, n_group "
+            f"({n_group})]"
+        )
```
[the remark this answers is not in the document]
Comment (time-machine2025) [reply flashinfer/fused_moe/fused_routing_dsv3.py:164]: Addressed in [ea92583](https://github.com/flashinfer-ai/flashinfer/commit/ea92583e25360fdeee8569489af920508ae3f135). The Python validator, native grouped launch gate, and test configuration filter now enforce `n_group <= 8`, `topk_group <= 4`, and at least two experts per group, while preserving the existing expert-count and selected-capacity limits. Native validation also checks positive group counts before division and enforces the top-k bounds when Python checks are bypassed.

The three grouped-limit regressions and the zero-top-k regression fail against the previous PR head and pass after this fix. The CPU validation file passes all 62 tests; both affected test files on L40S report 956 passed and 3960 skipped for unsupported parameter or architecture combinations. Could you re-check whether this addresses the finding?
